File: src/app/service-types.ts
import type { DepDirection } from "../domain/dep-tree";
import type { SkillOperationSummary, SkillTarget } from "../skills/types";
import type {
  EventRecord,
  Priority,
  RelationType,
  Task,
  TaskKind,
  TaskNote,
  TaskStatus,
} from "../types";

export interface InitResult {
  initialized: boolean;
  files: string[];
  skill_operation?: SkillOperationSummary;
}

export interface InitInput {
  installSkill?: boolean;
  uninstallSkill?: boolean;
  skillTargets?: SkillTarget[];
  skillName?: string;
  forceSkillOverwrite?: boolean;
  skillDirClaude?: string;
  skillDirCodex?: string;
  skillDirCopilot?: string;
  skillDirOpencode?: string;
}

export interface CreateInput {
  title: string;
  kind: TaskKind;
  priority: Priority;
  description?: string;
  externalRef?: string;
  parent?: string;
  exactId?: boolean;
}

export interface UpdateInput {
  id: string;
  title?: string;
  description?: string;
  clearDescription?: boolean;
  externalRef?: string;
  clearExternalRef?: boolean;
  status?: TaskStatus;
  priority?: Priority;
  exactId?: boolean;
}

export interface ClaimInput {
  id: string;
  assignee?: string;
  requireSpec?: boolean;
  exactId?: boolean;
}

export interface LinkInput {
  src: string;
  dst: string;
  type: RelationType;
  exactId?: boolean;
}

export interface DepInput {
  child: string;
  blocker: string;
  exactId?: boolean;
}

export interface SupersedeInput {
  source: string;
  withId: string;
  reason?: string;
  exactId?: boolean;
}

export interface DuplicateInput {
  source: string;
  canonical: string;
  reason?: string;
  exactId?: boolean;
}

export interface CloseInput {
  ids: string[];
  reason?: string;
  exactId?: boolean;
}

export interface ReopenInput {
  ids: string[];
  exactId?: boolean;
}

export interface HistoryInput {
  id: string;
  limit?: number;
  type?: string;
  actor?: string;
  since?: string;
  exactId?: boolean;
}

export interface HistoryResult {
  events: EventRecord[];
  count: number;
  truncated: boolean;
}

export interface DuplicateCandidatesResult {
  scanned: number;
  groups: Array<{
    key: string;
    tasks: Task[];
  }>;
}

export interface LabelInput {
  id: string;
  label: string;
  exactId?: boolean;
}

export interface DepTreeInput {
  id: string;
  direction?: DepDirection;
  depth?: number;
  exactId?: boolean;
}

export interface NoteAddInput {
  id: string;
  text: string;
  exactId?: boolean;
}

export interface NoteListInput {
  id: string;
  exactId?: boolean;
}

/** Input for attaching a markdown spec to a task. */
export interface SpecAttachInput {
  id: string;
  file?: string;
  source?: string;
  text?: string;
  stdin?: boolean;
  force?: boolean;
  exactId?: boolean;
}

export interface SpecCheckInput {
  id: string;
  exactId?: boolean;
}

export interface NoteAddResult {
  task_id: string;
  note: TaskNote;
  notes_count: number;
}

export interface NoteListResult {
  task_id: string;
  notes: TaskNote[];
}

export interface SpecAttachResult {
  task: Task;
  spec: {
    spec_path: string;
    spec_fingerprint: string;
    spec_attached_at: string;
    spec_attached_by: string;
    bytes: number;
  };
}

export type { SpecCheckDiagnostic, SpecCheckResult } from "./storage";
export type { SkillTarget } from "../skills/types";

export interface SearchInput {
  query: string;
}

/** Declarative filter applied to task listings. All fields are optional; unset fields match everything. */
export interface ListFilter {
  statuses?: TaskStatus[];
  assignee?: string;
  externalRef?: string;
  kind?: TaskKind;
  label?: string;
  labelAny?: string[];
  createdAfter?: string;
  updatedAfter?: string;
  closedAfter?: string;
  unassigned?: boolean;
  ids?: string[];
}

export interface StaleInput {
  days: number;
  status?: TaskStatus;
  assignee?: string;
}

export interface StaleResult {
  tasks: Task[];
  days: number;
  cutoff: string;
  statuses: TaskStatus[];
}

export interface DoctorResult {
  tasks: number;
  events: number;
  snapshot_loaded: boolean;
  warning?: string;
  issues: string[];
}

/** Constructor parameters shared by service method modules. */
export interface ServiceContext {
  repoRoot: string;
  actor: string;
  now: () => string;
}

File: src-rust/src/app/service_types.rs
pub use crate::app::storage::{SpecCheckDiagnostic, SpecCheckResult};
use crate::domain::dep_tree::DepDirection;
use crate::domain::validate::PlanningLane;
use crate::skills::types::SkillOperationSummary;
pub use crate::skills::types::SkillTarget;
use crate::types::{
    DependencyType, EventRecord, PlanningState, Priority, RelationType, RepairDep, Task, TaskKind,
    TaskNote, TaskStatus,
};
use serde::{Deserialize, Serialize};
use std::sync::Arc;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InitResult {
    pub initialized: bool,
    pub files: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub skill_operation: Option<SkillOperationSummary>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct InitInput {
    pub install_skill: bool,
    pub uninstall_skill: bool,
    pub skill_targets: Option<Vec<SkillTarget>>,
    pub skill_name: Option<String>,
    pub force_skill_overwrite: bool,
    pub skill_dir_claude: Option<String>,
    pub skill_dir_codex: Option<String>,
    pub skill_dir_copilot: Option<String>,
    pub skill_dir_opencode: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateInput {
    pub title: String,
    pub kind: TaskKind,
    pub priority: Priority,
    pub description: Option<String>,
    pub external_ref: Option<String>,
    pub discovered_from: Option<String>,
    pub parent: Option<String>,
    pub exact_id: bool,
    pub planning_state: Option<PlanningState>,
    pub explicit_id: Option<String>,
    pub body_file: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UpdateInput {
    pub id: String,
    pub title: Option<String>,
    pub description: Option<String>,
    pub clear_description: bool,
    pub external_ref: Option<String>,
    pub discovered_from: Option<String>,
    pub clear_discovered_from: bool,
    pub clear_external_ref: bool,
    pub status: Option<TaskStatus>,
    pub priority: Option<Priority>,
    pub exact_id: bool,
    pub planning_state: Option<PlanningState>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ClaimInput {
    pub id: String,
    pub assignee: Option<String>,
    pub require_spec: bool,
    pub exact_id: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LinkInput {
    pub src: String,
    pub dst: String,
    #[serde(rename = "type")]
    pub rel_type: RelationType,
    pub exact_id: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DepInput {
    pub child: String,
    pub blocker: String,
    pub dep_type: Option<DependencyType>,
    pub exact_id: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SupersedeInput {
    pub source: String,
    pub with_id: String,
    pub reason: Option<String>,
    pub exact_id: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DuplicateInput {
    pub source: String,
    pub canonical: String,
    pub reason: Option<String>,
    pub exact_id: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CloseInput {
    pub ids: Vec<String>,
    pub reason: Option<String>,
    pub exact_id: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ReopenInput {
    pub ids: Vec<String>,
    pub exact_id: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HistoryInput {
    pub id: String,
    pub limit: Option<usize>,
    #[serde(rename = "type")]
    pub event_type: Option<String>,
    pub actor: Option<String>,
    pub since: Option<String>,
    pub exact_id: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HistoryResult {
    pub events: Vec<EventRecord>,
    pub count: usize,
    pub truncated: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DuplicateCandidateGroup {
    pub key: String,
    pub tasks: Vec<Task>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DuplicateCandidatesResult {
    pub scanned: usize,
    pub groups: Vec<DuplicateCandidateGroup>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LabelInput {
    pub id: String,
    pub label: String,
    pub exact_id: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LabelCount {
    pub label: String,
    pub count: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DepTreeInput {
    pub id: String,
    pub direction: Option<DepDirection>,
    pub depth: Option<usize>,
    pub exact_id: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NoteAddInput {
    pub id: String,
    pub text: String,
    pub exact_id: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NoteListInput {
    pub id: String,
    pub exact_id: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SpecAttachInput {
    pub id: String,
    pub file: Option<String>,
    pub source: Option<String>,
    pub text: Option<String>,
    pub stdin: bool,
    pub force: bool,
    pub exact_id: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SpecCheckInput {
    pub id: String,
    pub exact_id: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NoteAddResult {
    pub task_id: String,
    pub note: TaskNote,
    pub notes_count: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NoteListResult {
    pub task_id: String,
    pub notes: Vec<TaskNote>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SpecAttachResult {
    pub task: Task,
    pub spec: SpecAttachSpec,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SpecAttachSpec {
    pub spec_path: String,
    pub spec_fingerprint: String,
    pub spec_attached_at: String,
    pub spec_attached_by: String,
    pub bytes: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SearchInput {
    pub query: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ListFilter {
    pub statuses: Option<Vec<TaskStatus>>,
    pub assignee: Option<String>,
    pub external_ref: Option<String>,
    pub discovered_from: Option<String>,
    pub kind: Option<TaskKind>,
    pub label: Option<String>,
    pub label_any: Option<Vec<String>>,
    pub created_after: Option<String>,
    pub updated_after: Option<String>,
    pub closed_after: Option<String>,
    pub unassigned: bool,
    pub ids: Option<Vec<String>>,
    pub planning_state: Option<PlanningState>,
    pub dep_type: Option<DependencyType>,
    pub dep_direction: Option<DepDirectionFilter>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DepDirectionFilter {
    In,
    Out,
    Any,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ReadyInput {
    pub lane: Option<PlanningLane>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MergeInput {
    pub sources: Vec<String>,
    pub into: String,
    pub reason: Option<String>,
    pub force: bool,
    pub dry_run: bool,
    pub exact_id: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MergeSummary {
    pub requested_sources: usize,
    pub merged_sources: usize,
    pub skipped_sources: usize,
    pub planned_events: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MergeProjected {
    pub target: Task,
    pub sources: Vec<Task>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MergeItem {
    pub id: String,
    pub status: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MergeTarget {
    pub id: String,
    pub title: String,
    pub status: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MergeResult {
    pub merged: Vec<MergeItem>,
    pub target: MergeTarget,
    pub dry_run: bool,
    pub warnings: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub plan_summary: Option<MergeSummary>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub projected: Option<MergeProjected>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StaleInput {
    pub days: i64,
    pub status: Option<TaskStatus>,
    pub assignee: Option<String>,
    pub limit: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StaleResult {
    pub tasks: Vec<Task>,
    pub days: i64,
    pub cutoff: String,
    pub statuses: Vec<TaskStatus>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DoctorResult {
    pub tasks: usize,
    pub events: usize,
    pub snapshot_loaded: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub warning: Option<String>,
    pub issues: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OrphanedLinkResult {
    pub src: String,
    pub dst: String,
    #[serde(rename = "type")]
    pub rel_type: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OrphansResult {
    pub orphaned_deps: Vec<RepairDep>,
    pub orphaned_links: Vec<OrphanedLinkResult>,
    pub total: usize,
}

#[derive(Clone)]
pub struct ServiceContext {
    pub repo_root: String,
    pub actor: String,
    pub now: Arc<dyn Fn() -> String + Send + Sync>,
}
